src/VirtoCommerce.MarketingModule.Data/Handlers/CouponUsageRecordHandler.cs

```diff
@@ -0,0 +1,88 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using VirtoCommerce.CoreModule.Core.Common;
+using VirtoCommerce.MarketingModule.Core.Model.Promotions;
+using VirtoCommerce.MarketingModule.Core.Model.Promotions.Search;
+using VirtoCommerce.MarketingModule.Core.Search;
+using VirtoCommerce.MarketingModule.Core.Services;
+using VirtoCommerce.OrdersModule.Core.Events;
+using VirtoCommerce.Platform.Core.Common;
+using VirtoCommerce.Platform.Core.Events;
+
+namespace VirtoCommerce.MarketingModule.Data.Handlers
+{
+
+
+    /// <summary>
+    /// Represents the logic of recording the use of coupons on both levels (Cart and Order).
+    /// </summary>
+    public class CouponUsageRecordHandler : IEventHandler<OrderChangedEvent>
+    {
+        private readonly IPromotionUsageService _usageService;
+        private readonly IPromotionUsageSearchService _promotionUsageSearchService;
+
+        public CouponUsageRecordHandler(IPromotionUsageService usageService, IPromotionUsageSearchService promotionUsageSearchService)
+        {
+            _usageService = usageService;
+            _promotionUsageSearchService = promotionUsageSearchService;
+            EqualityComparer = AnonymousComparer.Create((PromotionUsage x) => string.Join(":", x.PromotionId, x.CouponCode, x.ObjectId));
+        }
+
+        private IEqualityComparer<PromotionUsage> EqualityComparer { get; set; }
+
+        #region Implementation of IHandler<in OrderChangedEvent>
+
+        public async Task Handle(OrderChangedEvent message)
+        {
+            foreach (var changedEntry in message.ChangedEntries)
```

Comment:
I think better will be recording usages in background tasks so that not take influence for overall order creation time.